File: tourcount/src/main/java/com/wmstein/tourcount/database/CountDataSource.java
package com.wmstein.tourcount.database;

import android.annotation.SuppressLint;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;

import com.wmstein.tourcount.TourCountApplication;

import java.util.ArrayList;
import java.util.List;

import static com.wmstein.tourcount.database.DbHelper.COUNT_TABLE;
import static com.wmstein.tourcount.database.DbHelper.C_CODE;
import static com.wmstein.tourcount.database.DbHelper.C_COUNT_EI;
import static com.wmstein.tourcount.database.DbHelper.C_COUNT_F1I;
import static com.wmstein.tourcount.database.DbHelper.C_COUNT_F2I;
import static com.wmstein.tourcount.database.DbHelper.C_COUNT_F3I;
import static com.wmstein.tourcount.database.DbHelper.C_COUNT_LI;
import static com.wmstein.tourcount.database.DbHelper.C_COUNT_PI;

/********************************************************
 * Class CountDataSource provides methods for table Count
 * Basic structure by milo on 05/05/2014.
 * Created by wmstein on 2016-02-18,
 * last change on 2022-03-23
 */
public class CountDataSource
{
    // Database fields
    private SQLiteDatabase database;
    private final DbHelper dbHandler;
    private final String[] allColumns = {
        DbHelper.C_ID,
        C_COUNT_F1I,
        C_COUNT_F2I,
        C_COUNT_F3I,
        C_COUNT_PI,
        C_COUNT_LI,
        C_COUNT_EI,
        DbHelper.C_NAME,
        C_CODE,
        DbHelper.C_NOTES,
        DbHelper.C_NAME_G
    };

    private final TourCountApplication tourCountApp = new TourCountApplication();

    public CountDataSource(Context context)
    {
        dbHandler = new DbHelper(context);
    }

    public void open() throws SQLException
    {
        database = dbHandler.getWritableDatabase();
    }

    public void close()
    {
        dbHandler.close();
    }

    public void createCount(String name, String code, String name_g)
    {
        if (database.isOpen())
        {
            ContentValues values = new ContentValues();
            values.put(DbHelper.C_NAME, name);
            values.put(C_COUNT_F1I, 0);
            values.put(C_COUNT_F2I, 0);
            values.put(C_COUNT_F3I, 0);
            values.put(C_COUNT_PI, 0);
            values.put(C_COUNT_LI, 0);
            values.put(C_COUNT_EI, 0);
            values.put(C_CODE, code);
            values.put(DbHelper.C_NOTES, "");
            values.put(DbHelper.C_NAME_G, name_g);

            int insertId = (int) database.insert(COUNT_TABLE, null, values);
            Cursor cursor = database.query(COUNT_TABLE,
                allColumns, DbHelper.C_ID + " = " + insertId, null, null, null, null);
            cursor.close();
        }
    }

    @SuppressLint("Range")
    private Count cursorToCount(Cursor cursor)
    {
        Count newcount = new Count();
        newcount.id = cursor.getInt(cursor.getColumnIndex(DbHelper.C_ID));
        newcount.name = cursor.getString(cursor.getColumnIndex(DbHelper.C_NAME));
        newcount.count_f1i = cursor.getInt(cursor.getColumnIndex(C_COUNT_F1I));
        newcount.count_f2i = cursor.getInt(cursor.getColumnIndex(C_COUNT_F2I));
        newcount.count_f3i = cursor.getInt(cursor.getColumnIndex(C_COUNT_F3I));
        newcount.count_pi = cursor.getInt(cursor.getColumnIndex(C_COUNT_PI));
        newcount.count_li = cursor.getInt(cursor.getColumnIndex(C_COUNT_LI));
        newcount.count_ei = cursor.getInt(cursor.getColumnIndex(C_COUNT_EI));
        newcount.code = cursor.getString(cursor.getColumnIndex(C_CODE));
        newcount.notes = cursor.getString(cursor.getColumnIndex(DbHelper.C_NOTES));
        newcount.name_g = cursor.getString(cursor.getColumnIndex(DbHelper.C_NAME_G));
        return newcount;
    }

    public void deleteCountById(int id)
    {
        System.out.println("Gelöscht: Zähler mit ID: " + id);
        database.delete(COUNT_TABLE, DbHelper.C_ID + " = " + id, null);

    }

    public void saveCount(Count count)
    {
        ContentValues dataToInsert = new ContentValues();
        dataToInsert.put(C_COUNT_F1I, count.count_f1i);
        dataToInsert.put(C_COUNT_F2I, count.count_f2i);
        dataToInsert.put(C_COUNT_F3I, count.count_f3i);
        dataToInsert.put(C_COUNT_PI, count.count_pi);
        dataToInsert.put(C_COUNT_LI, count.count_li);
        dataToInsert.put(C_COUNT_EI, count.count_ei);
        dataToInsert.put(DbHelper.C_NAME, count.name);
        dataToInsert.put(C_CODE, count.code);
        dataToInsert.put(DbHelper.C_NOTES, count.notes);
        dataToInsert.put(DbHelper.C_NAME_G, count.name_g);
        String where = DbHelper.C_ID + " = ?";
        String[] whereArgs = {String.valueOf(count.id)};
        database.update(COUNT_TABLE, dataToInsert, where, whereArgs);
    }

    // save count_f1i
    public void saveCountf1i(Count count)
    {
        ContentValues dataToInsert = new ContentValues();
        dataToInsert.put(C_COUNT_F1I, count.count_f1i);
        String where = DbHelper.C_ID + " = ?";
        String[] whereArgs = {String.valueOf(count.id)};
        database.update(COUNT_TABLE, dataToInsert, where, whereArgs);
    }

    // save count_f2i
    public void saveCountf2i(Count count)
    {
        ContentValues dataToInsert = new ContentValues();
        dataToInsert.put(C_COUNT_F2I, count.count_f2i);
        String where = DbHelper.C_ID + " = ?";
        String[] whereArgs = {String.valueOf(count.id)};
        database.update(COUNT_TABLE, dataToInsert, where, whereArgs);
    }

    // save count_f3i
    public void saveCountf3i(Count count)
    {
        ContentValues dataToInsert = new ContentValues();
        dataToInsert.put(C_COUNT_F3I, count.count_f3i);
        String where = DbHelper.C_ID + " = ?";
        String[] whereArgs = {String.valueOf(count.id)};
        database.update(COUNT_TABLE, dataToInsert, where, whereArgs);
    }

    // save count_pi
    public void saveCountpi(Count count)
    {
        ContentValues dataToInsert = new ContentValues();
        dataToInsert.put(C_COUNT_PI, count.count_pi);
        String where = DbHelper.C_ID + " = ?";
        String[] whereArgs = {String.valueOf(count.id)};
        database.update(COUNT_TABLE, dataToInsert, where, whereArgs);
    }

    // save count_li
    public void saveCountli(Count count)
    {
        ContentValues dataToInsert = new ContentValues();
        dataToInsert.put(C_COUNT_LI, count.count_li);
        String where = DbHelper.C_ID + " = ?";
        String[] whereArgs = {String.valueOf(count.id)};
        database.update(COUNT_TABLE, dataToInsert, where, whereArgs);
    }

    // save count_ei
    public void saveCountei(Count count)
    {
        ContentValues dataToInsert = new ContentValues();
        dataToInsert.put(C_COUNT_EI, count.count_ei);
        String where = DbHelper.C_ID + " = ?";
        String[] whereArgs = {String.valueOf(count.id)};
        database.update(COUNT_TABLE, dataToInsert, where, whereArgs);
    }

    public void updateCountName(int id, String name, String code, String name_g)
    {
        if (database.isOpen())
        {
            ContentValues dataToInsert = new ContentValues();
            dataToInsert.put(DbHelper.C_NAME, name);
            dataToInsert.put(C_CODE, code);
            dataToInsert.put(DbHelper.C_NAME_G, name_g);
            String where = DbHelper.C_ID + " = ?";
            String[] whereArgs = {String.valueOf(id)};
            database.update(COUNT_TABLE, dataToInsert, where, whereArgs);
        }
    }

    public Count getCountById(int count_id)
    {
        Cursor cursor = database.query(COUNT_TABLE, allColumns,
            DbHelper.C_ID + " = " + count_id,
            null, null, null, null);

        cursor.moveToFirst();
        Count count = cursorToCount(cursor);
        cursor.close();
        return count;
    }

    // Used by CountingActivity
    public String[] getAllIds()
    {
        Cursor cursor = database.query(COUNT_TABLE, allColumns,
            null, null, null, null, null);

        String[] idArray = new String[cursor.getCount()];
        cursor.moveToFirst();
        int i = 0;
        while (!cursor.isAfterLast())
        {
            int uid = cursor.getInt(0);
            idArray[i] = Integer.toString(uid);
            i++;
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return idArray;
    }

    // Used by CountingActivity
    public String[] getAllIdsSrtName()
    {
        Cursor cursor = database.query(COUNT_TABLE, allColumns,
            null, null, null, null, DbHelper.C_NAME);

        String[] idArray = new String[cursor.getCount()];
        cursor.moveToFirst();
        int i = 0;
        while (!cursor.isAfterLast())
        {
            int uid = cursor.getInt(0);
            idArray[i] = Integer.toString(uid);
            i++;
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return idArray;
    }

    // Used by CountingActivity
    public String[] getAllIdsSrtCode()
    {
        Cursor cursor = database.query(COUNT_TABLE, allColumns,
            null, null, null, null, C_CODE);

        String[] idArray = new String[cursor.getCount()];
        cursor.moveToFirst();
        int i = 0;
        while (!cursor.isAfterLast())
        {
            int uid = cursor.getInt(0);
            idArray[i] = Integer.toString(uid);
            i++;
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return idArray;
    }

    // Used by CountingActivity
    public String[] getAllStrings(String sname)
    {
        Cursor cursor = database.query(COUNT_TABLE, allColumns,
            null, null, null, null, null);

        String[] uArray = new String[cursor.getCount()];
        int i = 0;
        cursor.moveToFirst();
        while (!cursor.isAfterLast())
        {
            @SuppressLint("Range") String uname = cursor.getString(cursor.getColumnIndex(sname));
            uArray[i] = uname;
            i++;
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return uArray;
    }

    // Used by CountingActivity
    public String[] getAllStringsSrtName(String sname)
    {

        Cursor cursor = database.query(COUNT_TABLE, allColumns,
            null, null, null, null, DbHelper.C_NAME);

        String[] uArray = new String[cursor.getCount()];

        cursor.moveToFirst();
        int i = 0;
        while (!cursor.isAfterLast())
        {
            @SuppressLint("Range") String uname = cursor.getString(cursor.getColumnIndex(sname));
            uArray[i] = uname;
            i++;
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return uArray;
    }

    // Used by CountingActivity
    public String[] getAllStringsSrtCode(String sname)
    {

        Cursor cursor = database.query(COUNT_TABLE, allColumns,
            null, null, null, null, C_CODE);

        String[] uArray = new String[cursor.getCount()];
        cursor.moveToFirst();
        int i = 0;
        while (!cursor.isAfterLast())
        {
            @SuppressLint("Range") String uname = cursor.getString(cursor.getColumnIndex(sname));
            uArray[i] = uname;
            i++;
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return uArray;
    }

    // Used by CountingActivity
    public Integer[] getAllImages()
    {
        Cursor cursor = database.query(COUNT_TABLE, allColumns,
            null, null, null, null, null);

        Integer[] imageArray = new Integer[cursor.getCount()];
        cursor.moveToFirst();
        int i = 0;
        while (!cursor.isAfterLast())
        {
            @SuppressLint("Range") String ucode = cursor.getString(cursor.getColumnIndex("code"));

            String rname = "p" + ucode; // species picture resource name
            int resId = tourCountApp.getResId(rname);
            int resId0 = tourCountApp.getResId("p00000");

            if (resId != 0)
            {
                imageArray[i] = resId;
            }
            else
            {
                imageArray[i] = resId0;
            }
            i++;
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return imageArray;
    }

    // Used by CountingActivity
    public Integer[] getAllImagesSrtName()
    {
        Cursor cursor = database.query(COUNT_TABLE, allColumns,
            null, null, null, null, DbHelper.C_NAME);

        Integer[] imageArray = new Integer[cursor.getCount()];
        cursor.moveToFirst();
        int i = 0;
        while (!cursor.isAfterLast())
        {
            @SuppressLint("Range") String ucode = cursor.getString(cursor.getColumnIndex("code"));

            String rname = "p" + ucode; // species picture resource name
            int resId = tourCountApp.getResId(rname);
            int resId0 = tourCountApp.getResId("p00000");

            if (resId != 0)
            {
                imageArray[i] = resId;
            }
            else
            {
                imageArray[i] = resId0;
            }
            i++;
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return imageArray;
    }

    // Used by CountingActivity
    public Integer[] getAllImagesSrtCode()
    {
        Cursor cursor = database.query(COUNT_TABLE, allColumns,
            null, null, null, null, C_CODE);

        Integer[] imageArray = new Integer[cursor.getCount()];
        cursor.moveToFirst();
        int i = 0;
        while (!cursor.isAfterLast())
        {
            @SuppressLint("Range") String ucode = cursor.getString(cursor.getColumnIndex("code"));

            String rname = "p" + ucode; // species picture resource name
            int resId = tourCountApp.getResId(rname);
            int resId0 = tourCountApp.getResId("p00000");

            if (resId != 0)
            {
                imageArray[i] = resId;
            }
            else
            {
                imageArray[i] = resId0;
            }
            i++;
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return imageArray;
    }

    // Used by EditSpecListActivity
    public List<Count> getAllSpecies()
    {
        List<Count> speci = new ArrayList<>();

        Cursor cursor = database.rawQuery("select * from " + COUNT_TABLE
            + " order by " + DbHelper.C_ID, null);

        cursor.moveToFirst();
        while (!cursor.isAfterLast())
        {
            Count count = cursorToCount(cursor);
            speci.add(count);
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return speci;
    }

    // Used by WelcomeActivity
    public int getDiffSpec()
    {
        int cntSpec = 0;
        Cursor cursor = database.rawQuery("select DISTINCT " + C_CODE + " from " + COUNT_TABLE
            + " WHERE " + " ("
            + C_COUNT_F1I + " > 0 or " + C_COUNT_F2I + " > 0 or "
            + C_COUNT_F3I + " > 0 or " + C_COUNT_PI + " > 0 or "
            + C_COUNT_LI + " > 0 or " + C_COUNT_EI + " > 0)", null);

        cursor.moveToFirst();
        while (!cursor.isAfterLast())
        {
            cntSpec++;
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return cntSpec;
    }

    // Used by EditSpecListActivity
    public List<Count> getAllSpeciesSrtName()
    {
        List<Count> speci = new ArrayList<>();

        Cursor cursor = database.rawQuery("select * from " + COUNT_TABLE
            + " order by " + DbHelper.C_NAME, null);

        cursor.moveToFirst();
        while (!cursor.isAfterLast())
        {
            Count count = cursorToCount(cursor);
            speci.add(count);
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return speci;
    }

    // Used by EditSpecListActivity and AddSpeciesActivity
    public List<Count> getAllSpeciesSrtCode()
    {
        List<Count> speci = new ArrayList<>();

        Cursor cursor = database.rawQuery("select * from " + COUNT_TABLE
            + " order by " + C_CODE, null);

        cursor.moveToFirst();
        while (!cursor.isAfterLast())
        {
            Count count = cursorToCount(cursor);
            speci.add(count);
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return speci;
    }

    // Used by ListSpeciesActivity
    public List<Count> getCntSpecies()
    {
        List<Count> speci = new ArrayList<>();

        Cursor cursor = database.rawQuery("select * from " + COUNT_TABLE
            + " WHERE ("
            + C_COUNT_F1I + " > 0 or " + C_COUNT_F2I + " > 0 or "
            + C_COUNT_F3I + " > 0 or " + C_COUNT_PI + " > 0 or "
            + C_COUNT_LI + " > 0 or " + C_COUNT_EI + " > 0)"
            + " order by " + DbHelper.C_ID, null, null);

        cursor.moveToFirst();
        while (!cursor.isAfterLast())
        {
            Count count = cursorToCount(cursor);
            speci.add(count);
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return speci;
    }

    // Used by ListSpeciesActivity
    public List<Count> getCntSpeciesSrtName()
    {
        List<Count> counts = new ArrayList<>();

        Cursor cursor = database.rawQuery("select * from " + COUNT_TABLE
            + " WHERE " + " ("
            + C_COUNT_F1I + " > 0 or " + C_COUNT_F2I + " > 0 or "
            + C_COUNT_F3I + " > 0 or " + C_COUNT_PI + " > 0 or "
            + C_COUNT_LI + " > 0 or " + C_COUNT_EI + " > 0)"
            + " order by " + DbHelper.C_NAME, null, null);

        cursor.moveToFirst();
        while (!cursor.isAfterLast())
        {
            Count count = cursorToCount(cursor);
            counts.add(count);
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return counts;
    }

    // Used by ListSpeciesActivity
    public List<Count> getCntSpeciesSrtCode()
    {
        List<Count> counts = new ArrayList<>();

        Cursor cursor = database.rawQuery("select * from " + COUNT_TABLE
            + " WHERE " + " ("
            + C_COUNT_F1I + " > 0 or " + C_COUNT_F2I + " > 0 or "
            + C_COUNT_F3I + " > 0 or " + C_COUNT_PI + " > 0 or "
            + C_COUNT_LI + " > 0 or " + C_COUNT_EI + " > 0)"
            + " order by " + C_CODE, null, null);

        cursor.moveToFirst();
        while (!cursor.isAfterLast())
        {
            Count count = cursorToCount(cursor);
            counts.add(count);
            cursor.moveToNext();
        }
        // Make sure to close the cursor
        cursor.close();
        return counts;
    }

}

File: tourcount/src/main/java/com/wmstein/tourcount/database/CountDataSource.kt
package com.wmstein.tourcount.database

import android.annotation.SuppressLint
import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import com.wmstein.tourcount.database.DbHelper.Companion.COUNT_TABLE

/********************************************************
 * Class CountDataSource provides methods for table Count
 * Basic structure by milo on 05/05/2014.
 * Created by wmstein on 2016-02-18,
 * last change on 2022-03-23,
 * converted to Kotlin on 2023-07-06,
 * last edited on 2025-03-15
 */
class CountDataSource(context: Context) {
    // Database fields
    private var database: SQLiteDatabase? = null
    private val dbHandler: DbHelper
    private val allColumns = arrayOf(
        DbHelper.C_ID,
        DbHelper.C_COUNT_F1I,
        DbHelper.C_COUNT_F2I,
        DbHelper.C_COUNT_F3I,
        DbHelper.C_COUNT_PI,
        DbHelper.C_COUNT_LI,
        DbHelper.C_COUNT_EI,
        DbHelper.C_NAME,
        DbHelper.C_CODE,
        DbHelper.C_NOTES,
        DbHelper.C_NAME_G
    )

    init {
        dbHandler = DbHelper(context)
    }

    @Throws(SQLException::class)
    fun open() {
        database = dbHandler.writableDatabase
    }

    fun close() {
        dbHandler.close()
    }

    // Used by AddSpeciesActivity
    fun createCount(name: String?, code: String?, name_g: String?) {
        if (database!!.isOpen) {
            val values = ContentValues()
            values.put(DbHelper.C_NAME, name)
            values.put(DbHelper.C_COUNT_F1I, 0)
            values.put(DbHelper.C_COUNT_F2I, 0)
            values.put(DbHelper.C_COUNT_F3I, 0)
            values.put(DbHelper.C_COUNT_PI, 0)
            values.put(DbHelper.C_COUNT_LI, 0)
            values.put(DbHelper.C_COUNT_EI, 0)
            values.put(DbHelper.C_CODE, code)
            values.put(DbHelper.C_NOTES, "")
            values.put(DbHelper.C_NAME_G, name_g)
            val insertId = database!!.insert(COUNT_TABLE, null, values).toInt()
            val cursor = database!!.query(COUNT_TABLE,
                allColumns, DbHelper.C_ID + " = " + insertId, null, null, null, null)
            cursor.close()
        }
    }

    @SuppressLint("Range")
    private fun cursorToCount(cursor: Cursor): Count {
        val newcount = Count()
        newcount.id = cursor.getInt(cursor.getColumnIndex(DbHelper.C_ID))
        newcount.name = cursor.getString(cursor.getColumnIndex(DbHelper.C_NAME))
        newcount.count_f1i = cursor.getInt(cursor.getColumnIndex(DbHelper.C_COUNT_F1I))
        newcount.count_f2i = cursor.getInt(cursor.getColumnIndex(DbHelper.C_COUNT_F2I))
        newcount.count_f3i = cursor.getInt(cursor.getColumnIndex(DbHelper.C_COUNT_F3I))
        newcount.count_pi = cursor.getInt(cursor.getColumnIndex(DbHelper.C_COUNT_PI))
        newcount.count_li = cursor.getInt(cursor.getColumnIndex(DbHelper.C_COUNT_LI))
        newcount.count_ei = cursor.getInt(cursor.getColumnIndex(DbHelper.C_COUNT_EI))
        newcount.code = cursor.getString(cursor.getColumnIndex(DbHelper.C_CODE))
        newcount.notes = cursor.getString(cursor.getColumnIndex(DbHelper.C_NOTES))
        newcount.name_g = cursor.getString(cursor.getColumnIndex(DbHelper.C_NAME_G))
        return newcount
    }

    fun deleteCountByCode(code: String) {
        println("Gelöscht: Zähler mit Code: $code")
        database!!.delete(COUNT_TABLE, DbHelper.C_CODE + " = '$code'", null)
    }

    fun saveCountNotes(count: Count) {
        val dataToInsert = ContentValues()
        dataToInsert.put(DbHelper.C_NOTES, count.notes)
        val where = DbHelper.C_ID + " = ?"
        val whereArgs = arrayOf(count.id.toString())
        database!!.update(COUNT_TABLE, dataToInsert, where, whereArgs)
    }

    // Save count_f1i
    fun saveCountf1i(count: Count) {
        val dataToInsert = ContentValues()
        dataToInsert.put(DbHelper.C_COUNT_F1I, count.count_f1i)
        val where = DbHelper.C_ID + " = ?"
        val whereArgs = arrayOf(count.id.toString())
        database!!.update(COUNT_TABLE, dataToInsert, where, whereArgs)
    }

    // Save count_f2i
    fun saveCountf2i(count: Count) {
        val dataToInsert = ContentValues()
        dataToInsert.put(DbHelper.C_COUNT_F2I, count.count_f2i)
        val where = DbHelper.C_ID + " = ?"
        val whereArgs = arrayOf(count.id.toString())
        database!!.update(COUNT_TABLE, dataToInsert, where, whereArgs)
    }

    // Save count_f3i
    fun saveCountf3i(count: Count) {
        val dataToInsert = ContentValues()
        dataToInsert.put(DbHelper.C_COUNT_F3I, count.count_f3i)
        val where = DbHelper.C_ID + " = ?"
        val whereArgs = arrayOf(count.id.toString())
        database!!.update(COUNT_TABLE, dataToInsert, where, whereArgs)
    }

    // Save count_pi
    fun saveCountpi(count: Count) {
        val dataToInsert = ContentValues()
        dataToInsert.put(DbHelper.C_COUNT_PI, count.count_pi)
        val where = DbHelper.C_ID + " = ?"
        val whereArgs = arrayOf(count.id.toString())
        database!!.update(COUNT_TABLE, dataToInsert, where, whereArgs)
    }

    // Save count_li
    fun saveCountli(count: Count) {
        val dataToInsert = ContentValues()
        dataToInsert.put(DbHelper.C_COUNT_LI, count.count_li)
        val where = DbHelper.C_ID + " = ?"
        val whereArgs = arrayOf(count.id.toString())
        database!!.update(COUNT_TABLE, dataToInsert, where, whereArgs)
    }

    // Save count_ei
    fun saveCountei(count: Count) {
        val dataToInsert = ContentValues()
        dataToInsert.put(DbHelper.C_COUNT_EI, count.count_ei)
        val where = DbHelper.C_ID + " = ?"
        val whereArgs = arrayOf(count.id.toString())
        database!!.update(COUNT_TABLE, dataToInsert, where, whereArgs)
    }

    // Used by EditSpecListActivity
    fun updateCountItem(id: Int, name: String?, code: String?, nameG: String?) {
        if (database!!.isOpen) {
            val dataToInsert = ContentValues()
            dataToInsert.put(DbHelper.C_NAME, name)
            dataToInsert.put(DbHelper.C_CODE, code)
            dataToInsert.put(DbHelper.C_NAME_G, nameG)
            val where = DbHelper.C_ID + " = ?"
            val whereArgs = arrayOf(id.toString())
            database!!.update(COUNT_TABLE, dataToInsert, where, whereArgs)
        }
    }

    // Used by WelcomeActivity
    fun writeCountItem(id: String?, code: String?, name: String?, nameG: String?) {
        if (database!!.isOpen) {
            val values = ContentValues()
            values.put(DbHelper.C_ID, id)
            values.put(DbHelper.C_NAME, name)
            values.put(DbHelper.C_COUNT_F1I, 0)
            values.put(DbHelper.C_COUNT_F2I, 0)
            values.put(DbHelper.C_COUNT_F3I, 0)
            values.put(DbHelper.C_COUNT_PI, 0)
            values.put(DbHelper.C_COUNT_LI, 0)
            values.put(DbHelper.C_COUNT_EI, 0)
            values.put(DbHelper.C_CODE, code)
            values.put(DbHelper.C_NOTES, "")
            values.put(DbHelper.C_NAME_G, nameG)
            database!!.insert(COUNT_TABLE, null, values)
        }
    }

    fun getCountById(count_id: Int): Count {
        val cursor = database!!.query(
            COUNT_TABLE, allColumns,
            DbHelper.C_ID + " = " + count_id,
            null, null, null, null
        )
        cursor.moveToFirst()
        val count = cursorToCount(cursor)
        cursor.close()
        return count
    }

    // Used by CountingActivity
    val allIds: Array<String?>
        get() {
            val cursor = database!!.query(
                COUNT_TABLE, allColumns,
                null, null, null, null, null
            )
            val idArray = arrayOfNulls<String>(cursor.count)
            cursor.moveToFirst()
            var i = 0
            while (!cursor.isAfterLast) {
                val uid = cursor.getInt(0)
                idArray[i] = uid.toString()
                i++
                cursor.moveToNext()
            }
            cursor.close()
            return idArray
        }

    // Used by CountingActivity
    val allIdsSrtName: Array<String?>
        get() {
            val cursor = database!!.query(
                COUNT_TABLE, allColumns,
                null, null, null, null, DbHelper.C_NAME
            )
            val idArray = arrayOfNulls<String>(cursor.count)
            cursor.moveToFirst()
            var i = 0
            while (!cursor.isAfterLast) {
                val uid = cursor.getInt(0)
                idArray[i] = uid.toString()
                i++
                cursor.moveToNext()
            }
            cursor.close()
            return idArray
        }

    // Used by CountingActivity
    val allIdsSrtCode: Array<String?>
        get() {
            val cursor = database!!.query(
                COUNT_TABLE, allColumns,
                null, null, null, null, DbHelper.C_CODE
            )
            val idArray = arrayOfNulls<String>(cursor.count)
            cursor.moveToFirst()
            var i = 0
            while (!cursor.isAfterLast) {
                val uid = cursor.getInt(0)
                idArray[i] = uid.toString()
                i++
                cursor.moveToNext()
            }
            cursor.close()
            return idArray
        }

    // Used by CountingActivity
    fun getAllStrings(sname: String?): Array<String?> {
        val cursor = database!!.query(
            COUNT_TABLE, allColumns,
            null, null, null, null, null
        )
        val uArray = arrayOfNulls<String>(cursor.count)
        var i = 0
        cursor.moveToFirst()
        while (!cursor.isAfterLast) {
            @SuppressLint("Range") val uname = cursor.getString(cursor.getColumnIndex(sname))
            uArray[i] = uname
            i++
            cursor.moveToNext()
        }
        cursor.close()
        return uArray
    }

    // Used by CountingActivity
    fun getAllStringsSrtName(sname: String?): Array<String?> {
        val cursor = database!!.query(
            COUNT_TABLE, allColumns,
            null, null, null, null, DbHelper.C_NAME
        )
        val uArray = arrayOfNulls<String>(cursor.count)
        cursor.moveToFirst()
        var i = 0
        while (!cursor.isAfterLast) {
            @SuppressLint("Range") val uname = cursor.getString(cursor.getColumnIndex(sname))
            uArray[i] = uname
            i++
            cursor.moveToNext()
        }
        cursor.close()
        return uArray
    }

    // Used by CountingActivity
    fun getAllStringsSrtCode(sname: String?): Array<String?> {
        val cursor = database!!.query(
            COUNT_TABLE, allColumns,
            null, null, null, null, DbHelper.C_CODE
        )
        val uArray = arrayOfNulls<String>(cursor.count)
        cursor.moveToFirst()
        var i = 0
        while (!cursor.isAfterLast) {
            @SuppressLint("Range") val uname = cursor.getString(cursor.getColumnIndex(sname))
            uArray[i] = uname
            i++
            cursor.moveToNext()
        }
        cursor.close()
        return uArray
    }

    // Used by EditSpecListActivity
    val allSpecies: List<Count>
        get() {
            val speci: MutableList<Count> = ArrayList()
            val cursor = database!!.rawQuery(
                "select * from " + COUNT_TABLE
                        + " order by " + DbHelper.C_ID, null
            )
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                val count = cursorToCount(cursor)
                speci.add(count)
                cursor.moveToNext()
            }
            cursor.close()
            return speci
        }

    // Used by WelcomeActivity
    val diffSpec: Int
        get() {
            var cntSpec = 0
            val cursor = database!!.rawQuery(
                "select DISTINCT " + DbHelper.C_CODE + " from " + COUNT_TABLE
                        + " WHERE " + " ("
                        + DbHelper.C_COUNT_F1I + " > 0 or " + DbHelper.C_COUNT_F2I + " > 0 or "
                        + DbHelper.C_COUNT_F3I + " > 0 or " + DbHelper.C_COUNT_PI + " > 0 or "
                        + DbHelper.C_COUNT_LI + " > 0 or " + DbHelper.C_COUNT_EI + " > 0)", null
            )
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                cntSpec++
                cursor.moveToNext()
            }
            cursor.close()
            return cntSpec
        }

    // Used by EditSpecListActivity
    val allSpeciesSrtName: List<Count>
        get() {
            val speci: MutableList<Count> = ArrayList()
            val cursor = database!!.rawQuery(
                "select * from " + COUNT_TABLE
                        + " order by " + DbHelper.C_NAME, null
            )
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                val count = cursorToCount(cursor)
                speci.add(count)
                cursor.moveToNext()
            }
            cursor.close()
            return speci
        }

    // Used by EditSpecListActivity and AddSpeciesActivity
    val allSpeciesSrtCode: List<Count>
        get() {
            val speci: MutableList<Count> = ArrayList()
            val cursor = database!!.rawQuery(
                "select * from " + COUNT_TABLE
                        + " order by " + DbHelper.C_CODE, null
            )
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                val count = cursorToCount(cursor)
                speci.add(count)
                cursor.moveToNext()
            }
            cursor.close()
            return speci
        }

    // Used by ShowResultsActivity
    val cntSpeciesSrtName: List<Count>
        get() {
            val counts: MutableList<Count> = ArrayList()
            val cursor = database!!.rawQuery(
                "select * from " + COUNT_TABLE
                        + " WHERE ("
                        + DbHelper.C_COUNT_F1I + " > 0 or " + DbHelper.C_COUNT_F2I + " > 0 or "
                        + DbHelper.C_COUNT_F3I + " > 0 or " + DbHelper.C_COUNT_PI + " > 0 or "
                        + DbHelper.C_COUNT_LI + " > 0 or " + DbHelper.C_COUNT_EI + " > 0)"
                        + " order by " + DbHelper.C_NAME, null, null
            )
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                val count = cursorToCount(cursor)
                counts.add(count)
                cursor.moveToNext()
            }
            cursor.close()
            return counts
        }

    // Used by ShowResultsActivity
    val cntSpeciesSrtCode: List<Count>
        get() {
            val counts: MutableList<Count> = ArrayList()
            val cursor = database!!.rawQuery(
                "select * from " + COUNT_TABLE
                        + " WHERE " + " ("
                        + DbHelper.C_COUNT_F1I + " > 0 or " + DbHelper.C_COUNT_F2I + " > 0 or "
                        + DbHelper.C_COUNT_F3I + " > 0 or " + DbHelper.C_COUNT_PI + " > 0 or "
                        + DbHelper.C_COUNT_LI + " > 0 or " + DbHelper.C_COUNT_EI + " > 0)"
                        + " order by " + DbHelper.C_CODE, null, null
            )
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                val count = cursorToCount(cursor)
                counts.add(count)
                cursor.moveToNext()
            }
            cursor.close()
            return counts
        }

    // Sorts COUNT_TABLE for C_CODE and contiguous index
    fun sortCounts() {
        var sql = "alter table 'counts' rename to 'counts_backup'"
        database!!.execSQL(sql)

        // create new counts table
        sql = ("create table counts("
                + DbHelper.C_ID + " integer primary key, "
                + DbHelper.C_COUNT_F1I + " int, "
                + DbHelper.C_COUNT_F2I + " int, "
                + DbHelper.C_COUNT_F3I + " int, "
                + DbHelper.C_COUNT_PI + " int, "
                + DbHelper.C_COUNT_LI + " int, "
                + DbHelper.C_COUNT_EI + " int, "
                + DbHelper.C_NAME + " text, "
                + DbHelper.C_CODE + " text, "
                + DbHelper.C_NOTES + " text, "
                + DbHelper.C_NAME_G + " text)")
        database!!.execSQL(sql)

        // insert the whole COUNT_TABLE data sorted into counts
        sql = ("INSERT INTO 'counts' (" +
                "'count_f1i', 'count_f2i', 'count_f3i', 'count_pi', 'count_li', 'count_ei', " +
                "'name', 'code', 'notes', 'name_g') " +
                "SELECT " +
                "count_f1i, count_f2i, count_f3i, count_pi, count_li, count_ei, " +
                "name, code, notes, name_g " +
                "from 'counts_backup' order by code")
        database!!.execSQL(sql)

        sql = "DROP TABLE 'counts_backup'"
        database!!.execSQL(sql)
    }

}
